src/integrationtests/java/com/aws/greengrass/integrationtests/logmanager/LogManagerTest.java

```diff
@@ -453,4 +461,174 @@ void GIVEN_system_config_with_small_periodic_interval_WHEN_active_logs_included_
+                    randomName1 + "UserComponentB.", randomName2);
+            logFiles.add(logFile);
+        }
+
+        setupKernel(tempDirectoryPath, "randomlyNamedFilesWithoutComponentsConfig.yaml");
+        //TODO: this lazy sleeping inherits from existing tests, but should be replaced by some better mechanism in
+        // future refactoring
+        TimeUnit.SECONDS.sleep(15);
+
+        // randomly rename file and see if uploader trying to process more file.
+        for (LogFile logFile : logFiles) {
+            String randomName3 = UUID.randomUUID().toString();
+            String randomName4 = UUID.randomUUID().toString();
+            LogFile renameFile = new LogFile(Files.createTempFile(tempDirectoryPath,
+                    randomName3 + "UserComponentB.", randomName4).toUri());
+            String tempFileHash = logFile.hashString();
+            logFile.renameTo(renameFile);
+            assertEquals(renameFile.hashString(), tempFileHash);
+            assertFalse(logFile.exists());
+        }
+        //TODO: this lazy sleeping inherits from existing tests, but should be replaced by some better mechanism in
+        // future refactoring
+        TimeUnit.SECONDS.sleep(15);
+
+        // Even the file are all renamed, the cloudWatchClient should only try to upload events once, and the total
+        // size should be contents of 4 files.
+        verify(cloudWatchLogsClient, times(1)).putLogEvents(captor.capture());
+
+        List<PutLogEventsRequest> putLogEventsRequests = captor.getAllValues();
+        assertEquals(1, putLogEventsRequests.size());
+        for (PutLogEventsRequest request : putLogEventsRequests) {
+            assertEquals(calculateLogStreamName(THING_NAME, LOCAL_DEPLOYMENT_GROUP_NAME), request.logStreamName());
+            assertEquals("/aws/greengrass/UserComponent/" + AWS_REGION + "/UserComponentB", request.logGroupName());
+            assertNotNull(request.logEvents());
+            assertEquals(DEFAULT_LOG_LINE_IN_FILE * testFileNumber, request.logEvents().size());
+            assertEquals(DEFAULT_FILE_SIZE * testFileNumber,
+                    request.logEvents().stream().mapToLong(value -> value.message().length()).sum());
+        }
+        // confirm no file is deleted
+        File folder = tempDirectoryPath.toFile();
+        Pattern logFileNamePattern = Pattern.compile("\\w*UserComponentB\\w*.\\w*");
+        List<File> allFiles = new ArrayList<>();
+        File[] files = folder.listFiles();
+        if (files != null) {
+            for (File file : files) {
+                if (file.isFile() && logFileNamePattern.matcher(file.getName()).find() && file.length() > 0) {
+                    allFiles.add(file);
+                }
+            }
+        }
+        assertEquals(testFileNumber, allFiles.size());
+    }
+
+    @Test
+    void GIVEN_files_randomly_named_WHEN_all_files_renamed_randomly_and_new_data_in_active_file_THEN_all_files_only_send_once()
```

Comment:
`new_data_in_active_file`, what about new data in some other file? In theory, we should be handling that too

Reply:
Are you referring to the scenario that new data can be written in rotated file? 
Our current logic will skip rotated and uploaded files when scanning.

Reply:
Yes. I don't love that, but OK.

Reply:
Agree. This will be included in future refactoring.